### ISSUE TITLE
"invalid parameter combination for AltiVec intrinsic" error with ppc64le, g++ v7.4

### ISSUE THREAD

**Issue by dncliss** [pytorch/pytorch#51592]
## 🐛 Bug

This is an independent follow-on issue (but somewhat related) to issue #50439, which was fixed by #50640 plus #51116.
When building pytorch on ppc64le, if using the g++ v7.4 compiler (but not v8), a number of `invalid parameter combination for AltiVec intrinsic` error is given.  These occur just a bit later than the aforementioned issue that was resolved; after passing that failure point successfully, this failure results.  (I should also mention that there was a different failure seen when using the v8 gcc/g++ compiler, only to emphasize that *this* is a separate unrelated issue; in fact, pytorch now builds successfully on ppc64le if updating to v8 compiler first). 

A number of the same type of error is reported, of this form:
```
../aten/src/ATen/cpu/vec256/vsx/vec256_complex_float_vsx.h:331:36: error: invalid parameter combination for AltiVec intrinsic __builtin_vec_vmrgew
     return {vec_mergee(_vec0, _vec0), vec_mergee(_vec1, _vec1)};
                                    ^
../aten/src/ATen/cpu/vec256/vsx/vec256_complex_float_vsx.h:331:62: error: invalid parameter combination for AltiVec intrinsic __builtin_vec_vmrgew
     return {vec_mergee(_vec0, _vec0), vec_mergee(_vec1, _vec1)};
```

The build command which preceded and produced the error was this:
```
FAILED: caffe2/CMakeFiles/vec256_test_all_types_VSX.dir/__/aten/src/ATen/test/vec256_test_all_types.cpp.o
/usr/bin/c++  -DCPU_CAPABILITY=VSX -DCPU_CAPABILITY_VSX -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DMAGMA_V2 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_EXTERNAL_MZCRC -D_FILE_OFFSET_BITS=64 -Iaten/src -I../aten/src -I. -I../ -I../cmake/../third_party/benchmark/include -Icaffe2/contrib/aten -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -I/usr/local/cuda/include -Iinclude -Icaffe2/../aten/src -Icaffe2/../aten/src/ATen -I../c10/.. -isystem third_party/gloo -isystem ../cmake/../third_party/gloo -isystem ../cmake/../third_party/googletest/googlemock/include -isystem ../cmake/../third_party/googletest/googletest/include -isystem ../third_party/protobuf/src -isystem ../cmake/../third_party/eigen -isystem /home/jenkins/miniconda/include/python3.7m -isystem /home/jenkins/miniconda/lib/python3.7/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem /usr/lib/powerpc64le-linux-gnu/openmpi/include/openmpi -isystem /usr/lib/powerpc64le-linux-gnu/openmpi/include/openmpi/opal/mca/event/libevent2022/libevent -isystem /usr/lib/powerpc64le-linux-gnu/openmpi/include/openmpi/opal/mca/event/libevent2022/libevent/include -isystem /usr/lib/powerpc64le-linux-gnu/openmpi/include -isystem ../cmake/../third_party/cub -isystem /usr/local/magma/include -isystem ../third_party/googletest/googletest/include -isystem ../third_party/googletest/googletest -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DNDEBUG -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow -DHAVE_VSX_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -fPIE   -DCAFFE2_USE_GLOO -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DTH_HAVE_THREAD -O3 -mcpu=power9 -mtune=power9 -mvsx -Wno-ignored-qualifiers -pthread -std=gnu++14 -MD -MT caffe2/CMakeFiles/vec256_test_all_types_VSX.dir/__/aten/src/ATen/test/vec256_test_all_types.cpp.o -MF caffe2/CMakeFiles/vec256_test_all_types_VSX.dir/__/aten/src/ATen/test/vec256_test_all_types.cpp.o.d -o caffe2/CMakeFiles/vec256_test_all_types_VSX.dir/__/aten/src/ATen/test/vec256_test_all_types.cpp.o -c ../aten/src/ATen/test/vec256_test_all_types.cpp
```

## To Reproduce

Well ... to reproduce, you'd need to run on ppc64le hardware, which I understand is not easily accessible.

I can point to the online public nightly executions on ppc64le in which you can examine the output, but I did switch the
main nightly build to use gcc v8, so it can't be referenced for this issue.  However, I have a separate test build set up on the
CI server that I can run at will (it's not run as a regular nightly build), and I can also execute things myself privately in a local container -- I can experiment as needed.

An example console output of the "invalid parameter combination" failure, with full context beyond just the error message reported above, can be observed here:
https://powerci.osuosl.org/job/pytorch-py3-test_gpu/65/console
(my suggestion is just to search on the words "invalid parameter combination").

## Expected behavior

Expected: Build should complete, whether using g++ v7 or v8.

## Environment

ppc64le with g++ 7.3 or 7.4, as noted above.  Build occurs on the master branch.

## Additional context

With apologies, to make an editorial/rationale comment:  Why not just mandate an update to v8?
As noted earlier, pytorch has built fine with v7 until the recent VSX feature, and if the feature were not to be supported with v7 it would be preferable to have it conditionally disabled based on compiler version (although it's still much better
if fixable to build with the feature and with v7).  For context, an example: There are other project environments that build pytorch collectively as just one of many bundled projects -- an example is `github.com/open-ce/open-ce` where the gcc/g++ 7 compiler is in use, as a collective build of pytorch and many other packages.

cc @malfet @seemethere @walterddr

**Comment by dncliss** [comment]
FYI @malfet and @quickwritereader -- if either of you have any thought on this one.

**Comment by malfet** [comment]
@dncliss can you please check, if setting BUILD_TEST to OFF and see if the build succeeds? If this is the case, probably easiest solution would be to disable the test if compiled with gcc7

**Comment by dncliss** [comment]
@malfet - sure, I'll try that.  I just added `export BUILD_TEST=OFF` to the run script and started it again, so we'll see; it's running now https://powerci.osuosl.org/job/pytorch-py3-test_gpu/66/console

**Comment by dncliss** [comment]
I see it failed the same way, seeming to have no effect.  Then again, although I don't have expertise in this, I am not sure how this plays into building tests, as it is not clear to me that building `caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/TensorCompareKernel.cpp.VSX.cpp.o` is specific to tests, and would be skipped if not building tests?  The failing command in full, from the above build console output, is:  

```
/usr/bin/c++ -DCPUINFO_SUPPORTED_PLATFORM=0 -DFMT_HEADER_ONLY=1 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DMAGMA_V2 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_DISTRIBUTED -DUSE_EXTERNAL_MZCRC -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -Dtorch_cpu_EXPORTS -Iaten/src -I../aten/src -I. -I../ -Icaffe2/contrib/aten -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -I/usr/local/cuda/include -I../torch/csrc/api -I../torch/csrc/api/include -I../caffe2/aten/src/TH -Icaffe2/aten/src/TH -Icaffe2/aten/src -Icaffe2/../aten/src -Icaffe2/../aten/src/ATen -I../torch/csrc -I../third_party/miniz-2.0.8 -I../aten/src/TH -I../aten/../third_party/catch/single_include -I../aten/src/ATen/.. -Icaffe2/aten/src/ATen -I../caffe2/core/nomnigraph/include -I../c10/.. -I../third_party/cpuinfo/include -I../third_party/FP16/include -I../third_party/tensorpipe -Ithird_party/tensorpipe -I../third_party/tensorpipe/third_party/libnop/include -I../third_party/fmt/include -isystem third_party/gloo -isystem ../cmake/../third_party/gloo -isystem ../third_party/protobuf/src -isystem ../cmake/../third_party/eigen -isystem /home/jenkins/miniconda/include/python3.7m -isystem /home/jenkins/miniconda/lib/python3.7/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem /usr/lib/powerpc64le-linux-gnu/openmpi/include/openmpi -isystem /usr/lib/powerpc64le-linux-gnu/openmpi/include/openmpi/opal/mca/event/libevent2022/libevent -isystem /usr/lib/powerpc64le-linux-gnu/openmpi/include/openmpi/opal/mca/event/libevent2022/libevent/include -isystem /usr/lib/powerpc64le-linux-gnu/openmpi/include -isystem ../cmake/../third_party/cub -isystem /usr/local/magma/include -isystem include -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DNDEBUG -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow -DHAVE_VSX_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -fPIC -DCAFFE2_USE_GLOO -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -fopenmp -DCAFFE2_BUILD_MAIN_LIB -pthread -std=gnu++14 -O3    -mcpu=power8 -mtune=power8 -mvsx -DCPU_CAPABILITY=VSX -DCPU_CAPABILITY_VSX -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/TensorCompareKernel.cpp.VSX.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/TensorCompareKernel.cpp.VSX.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/TensorCompareKernel.cpp.VSX.cpp.o -c aten/src/ATen/native/cpu/TensorCompareKernel.cpp.VSX.cpp
```

**Comment by malfet** [comment]
@dncliss is there a way I can trigger those builds (as right now my only way of experimenting is running snippets on https://godbolt.org/ )

**Comment by quickwritereader** [comment]
I just checked the workaround for it.
First one could be using aliased vector.
the second one is to switch to vec_perm.

let's try the first one.
I believe using _vecb0 instead of _vec0 pairs should be sufficient as it worked in godbolt.
` return {vec_mergee(_vecb0, _vecb0), vec_mergee(_vecb1, _vecb1)};`

**Comment by dncliss** [comment]
@malfet Good question.  I see that I was able to set the permission (at least temporarily) to allow an anonymous user to start the build. I could leave it like that temporarily but I don't want to allow anonymous users to edit stuff there.  But if you can only trigger the build without editing anything (as it just builds master), would that have any value?
(By this, I mean the test build at https://powerci.osuosl.org/job/pytorch-py3-test_gpu/ not the main nightly one).

A slightly more cumbersome but more official way would be to request free access at the OSUOSL lab (Oregon State University), through this service: https://osuosl.org/services/powerdev/ and trying to go through the "POWER CI Request" form.  In the short term though, otherwise, you can relay to me anything you'd like me to try.

I see the new comment from @quickwritereader here too, so I'll see what happens on my local build container with that suggestion (i.e. local private, you won't see in the viewable OSU lab build)

**Comment by dncliss** [comment]
With the _vecb0 instead of _vec0 change, replaced in each error location reported in `../aten/src/ATen/cpu/vec256/vsx/vec256_complex_float_vsx.h` , the reported "invalid parameter" error does go away, but a different error occurs. (In testing, I'm just running the isolated failing c++ command from above, not the entire build, just to see if it will compile `TensorCompareKernel.cpp.VSX.cpp.o` ).

There are two error messages in the output:
```
../aten/src/ATen/cpu/vec256/vsx/vsx_helpers.h:83:47: error: no matching function for call to 'vec_sldw_aux<3>(vbool32&, vbool32&)'
 #define vec_sldw(a, b, c) vec_sldw_aux<c>(a, b)
                                               ^
```
and
```
In file included from ../aten/src/ATen/cpu/vec256/vsx/vec256_common_vsx.h:14:0,
                 from ../aten/src/ATen/cpu/vec256/vec256.h:19,
                 from ../aten/src/ATen/native/cpu/Loops.h:35,
                 from aten/src/ATen/native/cpu/CopyKernel.cpp.VSX.cpp:6:
../aten/src/ATen/cpu/vec256/vsx/vec256_complex_float_vsx.h:297:81: error: could not convert '{<expression error>, <expression error>}' from '<brace-enclosed initializer list>' to 'at::vec256::{anonymous}::Vec256<c10::complex<float> >'
         vec_sldw(ret._vecb0, ret._vecb0, 3), vec_sldw(ret._vecb1, ret._vecb1, 3)};
                                                                                 ^
```
To produce this, the change to the header file `vec256_complex_float_vsx.h` was:
```
297c297
<         vec_sldw(ret._vec0, ret._vec0, 3), vec_sldw(ret._vec1, ret._vec1, 3)};
---
>         vec_sldw(ret._vecb0, ret._vecb0, 3), vec_sldw(ret._vecb1, ret._vecb1, 3)};
331c331
<     return {vec_mergee(_vec0, _vec0), vec_mergee(_vec1, _vec1)};
---
>     return {vec_mergee(_vecb0, _vecb0), vec_mergee(_vecb1, _vecb1)};
336c336
<     return {vec_mergeo(_vec0, _vec0), vec_mergeo(_vec1, _vec1)};
---
>     return {vec_mergeo(_vecb0, _vecb0), vec_mergeo(_vecb1, _vecb1)};
352,353c352,353
<         vec_mergee(first._vec0, second._vec0),
<         vec_mergee(first._vec1, second._vec1)};
---
>         vec_mergee(first._vecb0, second._vecb0),
>         vec_mergee(first._vecb1, second._vecb1)};
```

**Comment by quickwritereader** [comment]
I see. don't change vec_sldw. we have defined a function only for vector float for it

**Comment by dncliss** [comment]
Ahh!  I see, I over-corrected.  
So I put back vec_sldw on line 297, and the compile completed without error.  Whew!
Granted, this was a test of the one file that had failed rather than a full build, but it does look good.

**Comment by dncliss** [comment]
@quickwritereader  - I confirmed locally that the build worked fully to completion with your proposed fix above, so.... awaiting next steps toward getting this into master branch..... do you think this will go into a PR to make in time for 1.8.0 (I see that rc1 has been recently tagged)? Alternatively, it could be patched afterward, though it would be cleaner if it built upon release.

**Comment by dncliss** [comment]
If it would help, if it is an issue of workload or time, I'm willing to volunteer to submit the patch myself via a PR, since it is basically ready to go.  (My only reluctance is that I did not come up with the patch myself; I would not want to step on anyone's toes -- I simply tested it out, but full credit for the patch itself belongs to @quickwritereader as seen above).
 If you want me to handle that let me know -- I just would want to make clear not to be taking credit for the solution.
